### PR TITLE
Support contextual-biasing for streaming model

### PR DESCRIPTION
**Pull request by pkufool**

### CODE REVIEW

**Comment by pkufool** [comment]
Without context phrases:

```
2023-06-25 15:50:42.829755277 [E:onnxruntime:, env.cc:251 ThreadMain] pthread_setaffinity_np failed for thread: 2047229, index: 15, mask: {16, 52, }, error cod
e: 22 error msg: Invalid argument. Specify the number of threads explicitly so the affinity is not set.
2023-06-25 15:50:42.829757486 [E:onnxruntime:, env.cc:251 ThreadMain] pthread_setaffinity_np failed for thread: 2047230, index: 16, mask: {17, 53, }, error cod
e: 22 error msg: Invalid argument. Specify the number of threads explicitly so the affinity is not set.
Started!
Done!
exp-zh/test_wavs/3.wav
文森特卡索是全球知名的法国性格派演员
----------
exp-zh/test_wavs/4.wav
蒋友伯被拍到带着女儿出游
----------
exp-zh/test_wavs/5.wav
周望军就落实控股
----------
exp-zh/test_wavs/6.wav
朱莉南在上市见面会上表
----------
num_threads: 1
decoding_method: modified_beam_search
Wave duration: 20.328 s
Elapsed time: 4.031 s
Real time factor (RTF): 4.031/20.328 = 0.198
```

With context phrases:

```
2023-06-25 15:50:57.928258665 [E:onnxruntime:, env.cc:251 ThreadMain] pthread_setaffinity_np failed for thread: 2047405, index: 16, mask: {17, 53, }, error co$e: 22 error msg: Invalid argument. Specify the number of threads explicitly so the affinity is not set.
2023-06-25 15:50:57.928256533 [E:onnxruntime:, env.cc:251 ThreadMain] pthread_setaffinity_np failed for thread: 2047404, index: 15, mask: {16, 52, }, error co$e: 22 error msg: Invalid argument. Specify the number of threads explicitly so the affinity is not set.
Started!
Contexts list: ['文森特卡索', '蒋友博', '朱立楠', '周望君']
Done!
exp-zh/test_wavs/3.wav
文森特卡索是全球知名的法国性格派演员
----------
exp-zh/test_wavs/4.wav
蒋友博被拍到带着女儿出游
----------
exp-zh/test_wavs/5.wav
周望君就落实控股
----------
exp-zh/test_wavs/6.wav
朱立楠在上市见面会上表
----------
num_threads: 1
decoding_method: modified_beam_search
Wave duration: 20.328 s
Elapsed time: 3.886 s
Real time factor (RTF): 3.886/20.328 = 0.191
```

File: sherpa-onnx/csrc/online-transducer-decoder.h [hunk 1]
```diff
@@ -76,6 +78,27 @@ class OnlineTransducerDecoder {
...
+   * @param result  It is modified in-place.
+   *
+   * @note There is no need to pass encoder_out_length here since for the
+   * online decoding case, each utterance has the same number of frames
+   * and there are no paddings.
+   */
+  virtual void Decode(Ort::Value encoder_out, OnlineStream **ss,
+                      int32_t num_streams,
+                      std::vector<OnlineTransducerDecoderResult> *result) {
+    SHERPA_ONNX_LOGE("This interface is for ModifiedBeamSearchDecoder.");
```
Comment (csukuangfj) [inline sherpa-onnx/csrc/online-transducer-decoder.h]:
```suggestion
    SHERPA_ONNX_LOGE("This interface is for OnlineTransducerModifiedBeamSearchDecoder.");
```

File: sherpa-onnx/csrc/online-transducer-decoder.h [hunk 2]
```diff
@@ -76,6 +78,27 @@ class OnlineTransducerDecoder {
   virtual void Decode(Ort::Value encoder_out,
                       std::vector<OnlineTransducerDecoderResult> *result) = 0;
 
+  /** Run transducer beam search given the output from the encoder model.
+   *
+   * Note: Currently this interface is for contextual-biasing feature which
+   *       needs a ContextGraph owning by the OnlineStream.
```
Comment (csukuangfj) [inline sherpa-onnx/csrc/online-transducer-decoder.h]:
```suggestion
   *       needs a ContextGraph owned by the OnlineStream.
```

File: sherpa-onnx/csrc/online-transducer-decoder.h [hunk 3]
```diff
@@ -76,6 +78,27 @@ class OnlineTransducerDecoder {
...
+   *
+   * @param encoder_out A 3-D tensor of shape (N, T, joiner_dim)
+   * @param ss  A list of OnlineStreams.
+   * @param num_streams  The number of OnlineStreams, should be equal to N.
```
Comment (csukuangfj) [inline sherpa-onnx/csrc/online-transducer-decoder.h]: Do we really need this argument if it is always equal to N?

Also, when ss is nullptr, you set num_streams to 0, which is not equal to N.

File: sherpa-onnx/csrc/online-transducer-modified-beam-search-decoder.cc [hunk 4]
```diff
@@ -74,6 +81,9 @@ void OnlineTransducerModifiedBeamSearchDecoder::Decode(
   }
 
   int32_t batch_size = static_cast<int32_t>(encoder_out_shape[0]);
+
+  if (ss) SHERPA_ONNX_CHECK_EQ(batch_size, num_streams);
```
Comment (csukuangfj) [inline sherpa-onnx/csrc/online-transducer-modified-beam-search-decoder.cc]: We always use `{}` even for a single statement in sherpa-onnx.

**Review by csukuangfj — CHANGES_REQUESTED** [review]

File: sherpa-onnx/csrc/online-transducer-decoder.h (hunk 3 above)
Comment (pkufool) [reply sherpa-onnx/csrc/online-transducer-decoder.h]: I think we can remove it.

**Review by csukuangfj — APPROVED** [review]

**Pull request merged**